### ISSUE TITLE
calico hangs on `/proc/<pid>/cmdline` during `kubeadm reset -f`

### ISSUE THREAD

**Issue by microyahoo** [projectcalico/calico#7730]
## Expected Behavior

k8s cluster can be reset successfully.

## Current Behavior

calico hangs on `/proc/<pid>/cmdline` during `kubeadm reset -f`
1. `ps` and `top` will hang like below
```console
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=582, ...}) = 0
stat("/etc/localtime", {st_mode=S_IFREG|0644, st_size=582, ...}) = 0
write(1, "root      343250       2  0 22:5"..., 74root      343250       2  0 22:50 ?        00:00:00 [kworker/0:0-ata_sff]
) = 74
stat("/proc/344079", {st_mode=S_IFDIR|0555, st_size=0, ...}) = 0
openat(AT_FDCWD, "/proc/344079/stat", O_RDONLY) = 6
read(6, "344079 (calico-node) D 146808 34"..., 2048) = 303
close(6)                                = 0
openat(AT_FDCWD, "/proc/344079/status", O_RDONLY) = 6
read(6, "Name:\tcalico-node\nUmask:\t0022\nSt"..., 2048) = 1347
close(6)                                = 0
openat(AT_FDCWD, "/proc/344079/cmdline", O_RDONLY) = 6
read(6,
```
calico-node process is in D state
```console
~/go/src/ceph/build (test ✗) cat /proc/344079/status
Name:   calico-node
Umask:  0022
State:  D (disk sleep)
Tgid:   344079
Ngid:   0
Pid:    344079
PPid:   146808
TracerPid:      0
Uid:    0       0       0       0
Gid:    0       0       0       0
FDSize: 64
Groups:
NStgid: 344079  4042
NSpid:  344079  4042
NSpgid: 344079  4042
NSsid:  344079  4042
VmPeak:  1500320 kB
VmSize:  1436128 kB
VmLck:         0 kB
VmPin:         0 kB
VmHWM:     38164 kB
VmRSS:     38164 kB
RssAnon:           11376 kB
RssFile:           26788 kB
RssShmem:              0 kB
```
```console
~/go/src/ceph/build (test ✗) crictl ps -a
CONTAINER           IMAGE               CREATED             STATE               NAME                      ATTEMPT             POD ID
4ef22af44a4db       e7f3bebbeed56       7 hours ago         Exited              chrony-sync               0                   3bae15f10bbdf
2c9f9126dee6b       bf119bc65a556       7 hours ago         Exited              postgres                  0                   3559a202801fa
fa47ebdfb0fec       5d3d5ddc8605d       7 hours ago         Exited              calico-kube-controllers   0                   3ef1ff1dba706
6c121c03e038f       8d147537fb7d1       7 hours ago         Exited              coredns                   0                   c11c85f5841ce
f959bcc817850       8d147537fb7d1       7 hours ago         Exited              coredns                   0                   96f3d5f4ee55f
6140064de4586       c4d75af7e098e       7 hours ago         Running             calico-node               0                   482dad02b37e0
d2598cf4a2991       5660150975fb8       7 hours ago         Exited              flexvol-driver            0                   482dad02b37e0
a84d72e3e5b4e       5749e8b276f9b       7 hours ago         Exited              install-cni               0                   482dad02b37e0
688f72725ba0f       5749e8b276f9b       7 hours ago         Exited              upgrade-ipam              0                   482dad02b37e0
```
Partial output after execution of `kubedam reset -f`
```console
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.336396 1072011 etcd.go:102] etcd endpoints read from pods: https://10.9.9.150:2379,h
ttps://10.9.9.151:2379,https://10.9.9.152:2379
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.352055 1072011 etcd.go:222] etcd endpoints read from etcd: https://10.9.9.150:2379,h
ttps://10.9.9.151:2379
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.352080 1072011 etcd.go:120] update etcd endpoints: https://10.9.9.150:2379,https://1
0.9.9.151:2379
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.363032 1072011 local.go:118] [etcd] get the member id from peer: https://10.9.9.150:
2380
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.375987 1072011 local.go:124] [etcd] removing etcd member: https://10.9.9.150:2380, i
d: 984746417049005223
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.415411 1072011 local.go:129] [etcd] Updated etcd member list: [{g2 https://10.9.9.15
1:2380}]
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.415462 1072011 cleanupnode.go:60] [reset] Getting init system
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] [reset] Stopping the kubelet service
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] [reset] Unmounting mounted directories in "/var/lib/kubelet"
17:28:39 [INFO] [ssh.go:65] [10.9.9.150:22] I0601 17:28:39.532355 1072011 cleanupnode.go:82] [reset] Removing Kubernetes-managed containers
```

## Steps to Reproduce (for bugs)

I don't know how to reproduce it, but it's 100% present in my system when invoking `kubeadm reset -f`.

## Your Environment

* Calico version: `docker.io/calico/node:v3.19.1`
* Orchestrator version (e.g. kubernetes, mesos, rkt): `v1.22.7`
* Operating System and version: 
```console
~/go/src/ceph/build (test ✗) uname -r
4.18.0-490.el8.x86_64
~/go/src/ceph/build (test ✗) cat /etc/redhat-release
CentOS Stream release 8
```
* Link to your project (optional):

**Comment by tomastigera** [comment]
What is supposed to work during such a reset? It does not sound like and unexpected behaviour that things break when you reset stuff under their hands :thinking:

**Comment by microyahoo** [comment]
I think calico should be reset normally, furthermore, I am not sure whether it is related to virtual machine.
```console
 3042 May 29 15:35:25 g1 kernel: INFO: task calico-node:2982389 blocked for more than 120 seconds.        
 3043 May 29 15:35:25 g1 kernel:      Tainted: G           OE    --------- -  - 4.18.0-490.el8.x86_64 #1  
 3044 May 29 15:35:25 g1 kernel: "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
 3045 May 29 15:35:25 g1 kernel: task:calico-node     state:D stack:    0 pid:2982389 ppid:2022278 flags:0x00000084
 3046 May 29 15:35:25 g1 kernel: Call Trace:                                                                                                                                                       
 3047 May 29 15:35:25 g1 kernel: __schedule+0x2d1/0x870                                                   
 3048 May 29 15:35:25 g1 kernel: ? smp_call_function_single+0xc8/0xf0                                     
 3049 May 29 15:35:25 g1 kernel: schedule+0x55/0xf0                                                       
 3050 May 29 15:35:25 g1 kernel: io_schedule+0x12/0x40                                                    
 3051 May 29 15:35:25 g1 kernel: migration_entry_wait_on_locked+0x1ea/0x290                               
 3052 May 29 15:35:25 g1 kernel: ? filemap_fdatawait_keep_errors+0x50/0x50                                
 3053 May 29 15:35:25 g1 kernel: do_swap_page+0x5b0/0x710                                                 
 3054 May 29 15:35:25 g1 kernel: ? pmd_devmap_trans_unstable+0x2e/0x40                                                                                                                             
 3055 May 29 15:35:25 g1 kernel: ? handle_pte_fault+0x5d/0x880                                            
 3056 May 29 15:35:25 g1 kernel: __handle_mm_fault+0x365/0x6c0                                            
 3057 May 29 15:35:25 g1 kernel: handle_mm_fault+0xca/0x2a0                                               
 3058 May 29 15:35:25 g1 kernel: __do_page_fault+0x1f0/0x450                                              
 3059 May 29 15:35:25 g1 kernel: ? _copy_to_user+0x22/0x40                                                
 3060 May 29 15:35:25 g1 kernel: do_page_fault+0x37/0x130                                                 
 3061 May 29 15:35:25 g1 kernel: ? page_fault+0x8/0x30                                                    
 3062 May 29 15:35:25 g1 kernel: page_fault+0x1e/0x30                                                     
 3063 May 29 15:35:25 g1 kernel: RIP: 0033:0x4247f2                                                       
 3064 May 29 15:35:25 g1 kernel: Code: Unable to access opcode bytes at RIP 0x4247c8.                     
 3065 May 29 15:35:25 g1 kernel: RSP: 002b:00007ffda3374480 EFLAGS: 00010293                              
 3066 May 29 15:35:25 g1 kernel: RAX: 0000000000203000 RBX: 00007fd769360c22 RCX: 0000000000000000                                                                                                 
 3067 May 29 15:35:25 g1 kernel: RDX: 0000000000000020 RSI: 0000000000000000 RDI: 000000c0009f8440        
 3068 May 29 15:35:25 g1 kernel: RBP: 00007ffda3374500 R08: 00007fd769510fff R09: 000000c0009f8440        
 3069 May 29 15:35:25 g1 kernel: R10: 000000c00005d698 R11: 0000000000000000 R12: 0000000000000011        
 3070 May 29 15:35:25 g1 kernel: R13: 0000000000000009 R14: 0000000002cd08a6 R15: 0000000000000039 
```

**Comment by veinito** [comment]
I observe exactly the same thing on my RHEL 8.8 VM with following setup:
- calico version: 3.13.3
- kubernetes: RKE2 1.20 (containerd) or RKE1 (docker 20.10)

I don't call any `kubeadm reset` commands... the issues just happens every time after 3-10 since the system reboot

**Comment by dlafcracy** [comment]
Getting similar issue where calico became defunct / zombie process. But I didnt do kubeadm reset too.

Calico Version: 3.19.3
Kubernetes server Version: v1.20.11 (containerd)
OS: RHEL 8.8 (running kernel 4.18.0-477.10.1.el8_8.x86_64)

After OS rebooted, from time to time across different nodes in the cluster, I can see OS messages reporting the calico-node task blocked error similar to @microyahoo 

When I tried to do a strace ps -ef (because I kept getting hung up on ps -ef command), it is also stuck on reading /proc/<pid>/cmdline where pid is for calico-node or calico-ipam.
If I tried to just cat /proc/<pid>/cmdline, it will also hung up.

The only bypass I have successfully did was, revert to older OS kernel (tested RHEL 8.7 4.18.0-425.3.1.el8.x86_64, and RHEL 8.6 4.18.0-372.41.1.el8_6.x86_64) and then the issue stopped happening.
